File: server/ucp-api/src/adapters/outbound/identity/zitadel-projects.adapter.ts
import { Injectable } from '@nestjs/common';
import { ZitadelBaseClient } from './zitadel-base.client';
import { IProjectProvider } from '../../../ports/outbound/project.provider';
import {
  ZitadelRolesResponseSchema,
  ZitadelProjectGrantsResponseSchema,
  ZitadelRawUserSearchResponseSchema,
} from '../../../domain/dtos/zitadel.dto';

@Injectable()
export class ZitadelProjectsAdapter
  extends ZitadelBaseClient
  implements IProjectProvider
{
  async createProjectGrant(
    orgId: string,
    projectId: string,
    roleKeys: string[],
  ): Promise<void> {
    this.logger.log(
      `Creating project grant in Zitadel. OrgId: ${orgId}, ProjectId: ${projectId}`,
    );

    try {
      const grantResponse = await this.fetchWithAuth(
        `/management/v1/projects/${projectId}/grants`,
        {
          method: 'POST',
          body: JSON.stringify({
            grantedOrgId: orgId,
            roleKeys: roleKeys,
          }),
        },
      );

      if (!grantResponse.ok)
        await this.handleResponseError(grantResponse, 'create project grant');

      this.logger.log(
        `Successfully granted Project ${projectId} to Organization ${orgId}`,
      );
    } catch (error) {
      this.logger.error(`Error creating project grant for org ${orgId}`, error);
      throw error;
    }
  }

  async deleteProjectGrant(
    orgId: string,
    projectId: string,
  ): Promise<void> {
    this.logger.log(
      `Deleting project grant in Zitadel. OrgId: ${orgId}, ProjectId: ${projectId}`,
    );

    try {
      // First, search for the grant to get its ID
      const searchResponse = await this.fetchWithAuth(
        `/management/v1/projects/${projectId}/grants/_search`,
        {
          method: 'POST',
          body: JSON.stringify({ queries: [] }),
        },
      );

      if (!searchResponse.ok)
        await this.handleResponseError(searchResponse, 'search project grants');

      const searchData = (await searchResponse.json()) as unknown;
      const parsedSearchData = ZitadelProjectGrantsResponseSchema.parse(searchData);
      const grant = parsedSearchData.result?.find((g: any) => g.grantedOrgId === orgId);

      if (!grant) {
        this.logger.warn(
          `No project grant found for org ${orgId} and project ${projectId}. Skipping deletion.`,
        );
        return;
      }

      const grantId = grant.grantId || grant.id;

      // Delete the grant
      const deleteResponse = await this.fetchWithAuth(
        `/management/v1/projects/${projectId}/grants/${grantId}`,
        {
          method: 'DELETE',
        },
      );

      if (!deleteResponse.ok)
        await this.handleResponseError(deleteResponse, 'delete project grant');

      this.logger.log(
        `Successfully revoked Project ${projectId} from Organization ${orgId}`,
      );
    } catch (error) {
      this.logger.error(`Error deleting project grant for org ${orgId}`, error);
      throw error;
    }
  }

  async getRoles(): Promise<
    import('../../../domain/dtos/zitadel.dto').ZitadelRole[]
  > {
    this.logger.log('Fetching roles for UCP Project');

    const response = await this.fetchWithAuth(
      `/management/v1/projects/${this.ucpProjectId}/roles/_search`,
      {
        method: 'POST',
        body: JSON.stringify({}),
      },
    );

    if (!response.ok) await this.handleResponseError(response, 'fetch roles');

    const data = (await response.json()) as unknown;
    const parsedData = ZitadelRolesResponseSchema.parse(data);
    return parsedData.result || [];
  }

  async getUsers(
    orgId: string,
  ): Promise<import('../../../domain/dtos/zitadel.dto').ZitadelUser[]> {
    this.logger.log(`Fetching users for org ${orgId}`);

    // 1. Fetch all users in the org
    const response = await this.fetchWithAuth('/management/v1/users/_search', {
      method: 'POST',
      headers: {
        'x-zitadel-orgid': orgId,
      },
      body: JSON.stringify({}),
    });

    if (!response.ok) await this.handleResponseError(response, 'fetch users');

    const data = (await response.json()) as unknown;
    const rawUsersData = ZitadelRawUserSearchResponseSchema.parse(data);
    const users = rawUsersData.result;

    // 2. Fetch grants for each user individually
    const usersWithRoles = await Promise.all(
      users.map(async (u: any) => {
        let role = 'Unknown';
        try {
          const grantRes = await this.fetchWithAuth(
            '/management/v1/users/grants/_search',
            {
              method: 'POST',
              headers: {
                'x-zitadel-orgid': orgId,
              },
              body: JSON.stringify({
                queries: [{ userIdQuery: { userId: u.id } }],
              }),
            },
          );
          if (grantRes.ok) {
            const grantData = (await grantRes.json()) as unknown;
            const parsedGrantData =
              ZitadelProjectGrantsResponseSchema.parse(grantData);
            const grant = parsedGrantData.result?.find(
              (g: any) => g.projectId === this.ucpProjectId,
            );
            if (grant?.roleKeys?.length) {
              role = grant.roleKeys[0];
            } else {
              role = 'Unknown';
            }
          }
        } catch {
          // leave role as Unknown if grant fetch fails
        }

        return {
          id: u.id,
          email: u.human?.email?.email || u.userName,
          displayName: u.human?.profile?.displayName || u.userName,
          firstName: u.human?.profile?.firstName,
          lastName: u.human?.profile?.lastName,
          state: u.state,
          role,
          createdAt: u.details?.creationDate,
        };
      }),
    );

    return usersWithRoles;
  }
}

File: core/ucp/packages/ucp/src/ucp/adapters/outbound/identity/zitadel_projects_adapter.py
import structlog

from ucp.adapters.outbound.identity.zitadel_client import ZitadelClient
from ucp.domain.dtos.zitadel_dtos import (
    ZitadelProjectGrantsResponse,
    ZitadelRawUserSearchResponse,
    ZitadelRole,
    ZitadelRolesResponse,
    ZitadelUser,
)
from ucp.ports.outbound.project_provider_port import ProjectProviderPort

logger = structlog.get_logger(__name__)


class ZitadelProjectsAdapter(ZitadelClient, ProjectProviderPort):
    async def create_project_grant(
        self, org_id: str, project_id: str, role_keys: list[str]
    ) -> None:
        logger.info(
            "Creating project grant in Zitadel. OrgId: {org_id}, ProjectId: {project_id}",
            org_id=org_id,
            project_id=project_id,
        )

        try:
            response = await self.fetch_with_auth(
                endpoint=f"/management/v1/projects/{project_id}/grants",
                method="POST",
                json={
                    "grantedOrgId": org_id,
                    "roleKeys": role_keys,
                },
            )

            if response.status_code >= 400:
                await self.handle_response_error(response, "create project grant")

            logger.info(
                "Successfully granted Project {project_id} to Organization {org_id}",
                project_id=project_id,
                org_id=org_id,
            )
        except Exception:
            logger.exception("Error creating project grant for org {org_id}", org_id=org_id)

            raise

    async def delete_project_grant(self, org_id: str, project_id: str) -> None:
        logger.info(
            "Deleting project grant in Zitadel. OrgId: {org_id}, ProjectId: {project_id}",
            org_id=org_id,
            project_id=project_id,
        )

        try:
            # First, search for the grant to get its ID
            search_response = await self.fetch_with_auth(
                endpoint=f"/management/v1/projects/{project_id}/grants/_search",
                method="POST",
                json={"queries": []},
            )

            if search_response.status_code >= 400:
                await self.handle_response_error(search_response, "search project grants")

            search_data = search_response.json()
            parsed_search_data = ZitadelProjectGrantsResponse.model_validate(search_data)

            grant = next((g for g in parsed_search_data.result if g.granted_org_id == org_id), None)

            if not grant:
                logger.warning(
                    "No project grant found for org {org_id} and project {project_id}. Skipping deletion.",
                    org_id=org_id,
                    project_id=project_id,
                )
                return

            grant_id = grant.grant_id or grant.id

            # Delete the grant
            delete_response = await self.fetch_with_auth(
                endpoint=f"/management/v1/projects/{project_id}/grants/{grant_id}", method="DELETE"
            )

            if delete_response.status_code >= 400:
                await self.handle_response_error(delete_response, "delete project grant")

            logger.info(
                "Successfully revoked Project {project_id} from Organization {org_id}",
                project_id=project_id,
                org_id=org_id,
            )
        except Exception:
            logger.exception("Error deleting project grant for org {org_id}", org_id=org_id)

            raise

    async def get_roles(self) -> list[ZitadelRole]:
        logger.info("Fetching roles for UCP Project")

        response = await self.fetch_with_auth(
            endpoint=f"/management/v1/projects/{self.ucp_project_id}/roles/_search",
            method="POST",
            json={},
        )

        if response.status_code >= 400:
            await self.handle_response_error(response, "fetch roles")

        data = response.json()
        parsed_data = ZitadelRolesResponse.model_validate(data)
        return parsed_data.result

    async def get_users(self, org_id: str) -> list[ZitadelUser]:
        logger.info("Fetching users for org {org_id}", org_id=org_id)

        # 1. Fetch all users in the org
        response = await self.fetch_with_auth(
            endpoint="/management/v1/users/_search",
            method="POST",
            headers={"x-zitadel-orgid": org_id},
            json={},
        )

        if response.status_code >= 400:
            await self.handle_response_error(response, "fetch users")

        data = response.json()
        raw_users_data = ZitadelRawUserSearchResponse.model_validate(data)
        users = raw_users_data.result

        # 2. Fetch all grants for the UCP Project in this Org
        grants = []
        try:
            grant_res = await self.fetch_with_auth(
                endpoint=f"/management/v1/projects/{self.ucp_project_id}/grants/_search",
                method="POST",
                json={"queries": []},
            )
            if grant_res.status_code < 400:
                grant_data = grant_res.json()
                parsed_grant_data = ZitadelProjectGrantsResponse.model_validate(grant_data)
                grants = parsed_grant_data.result
        except Exception:  # noqa: BLE001 - grant fetch is non-critical; empty grants is a safe fallback
            logger.warning("Failed to fetch grants for org %s", org_id)

        # 3. Map grants to users in memory
        org_grant = next((g for g in grants if g.granted_org_id == org_id), None)
        role = org_grant.role_keys[0] if org_grant and org_grant.role_keys else "Unknown"

        users_with_roles = []
        for u in users:
            email = u.human.email.email if u.human and u.human.email else u.user_name
            display_name = (
                u.human.profile.display_name if u.human and u.human.profile else u.user_name
            )
            first_name = u.human.profile.first_name if u.human and u.human.profile else None
            last_name = u.human.profile.last_name if u.human and u.human.profile else None

            users_with_roles.append(
                ZitadelUser(
                    userId=u.id,
                    email=email,
                    displayName=display_name,
                    firstName=first_name,
                    lastName=last_name,
                    state=u.state,
                    role=role,
                    createdAt=u.details.creation_date if u.details else None,
                )
            )

        return users_with_roles
